R/methods.R

```diff
@@ -81,58 +96,61 @@ NULL
 ## projections. For each projection, it evaluates the fun-function, which
 ## calculates the linear predictor if called from proj_linpred and samples from
 ## the predictive distribution if called from proj_predict.
-proj_helper <- function(object, newdata, offsetnew, weightsnew, nterms, seed,
-                        proj_predict, ...) {
+proj_helper <- function(object, filter_nterms = NULL, newdata,
+                        offsetnew, weightsnew,
+                        onesub_fun, integrated = NULL, transform = NULL,
+                        size_sub = NULL, ...) {
   if (inherits(object, "projection") ||
-    (length(object) > 0 && inherits(object[[1]], "projection"))) {
-    proj <- object
+      (length(object) > 0 && inherits(object[[1]], "projection"))) {
+    if (!is.null(filter_nterms)) {
+      if (!.is_proj_list(object)) {
+        object <- list(object)
+      }
+      projs <- Filter(function(x) {
+        count_terms_chosen(x$solution_terms) %in% (filter_nterms + 1)
+      }, object)
+    } else {
+      projs <- object
+    }
   } else {
     ## reference model or varsel object obtained, so run the projection
-    proj <- project(object = object, nterms = nterms, seed = seed, ...)
+    projs <- project(object = object, ...)
   }
 
-  if (!.is_proj_list(proj)) {
-    proj <- list(proj)
+  if (!.is_proj_list(projs)) {
+    projs <- list(projs)
   } else {
-    ## proj is not a projection object
-    if (any(sapply(proj, function(x) !("family" %in% names(x))))) {
-      stop(paste(
-        "proj_linpred only works with objects returned by",
-        " varsel, cv_varsel or project"
-      ))
+    ## projs is some other object, not containing an element called "family" (so
+    ## it could be a `proj_list` but must not necessarily)
+    if (any(sapply(projs, function(x) !("family" %in% names(x))))) {
+      stop("Invalid object supplied to argument `object`.")
     }
   }
 
   if (is.null(newdata)) {
     ## pick first projection's function
-    newdata <- proj[[1]]$refmodel$fetch_data()
```

Comment:
I'm wondering, can we create a new structure `projection_list` that has all common slots as attributes? That seems a way better way of not duplicating all the internal structures. We should anyway make sure that when using any of the individual projections we have all the required slots available. I'll think on this.

Reply:
Yes, probably a good idea. Or simply always force a list (which is of length one in the case of a single submodel).

Reply:
In any case, that's a larger change, I guess. So perhaps it should be performed outside of this PR.